Split upload task to allow tracing changed files with dry run

justin-thurman opened the pull request:
# Description

This PR addresses a problem where it was not possible to trace files affected by recent changes (via `--trace-changed`) when doing a dry run (`--dry-run`). This is because the `traceChangedFiles` step was part of the `upload` task, which was skipped on dry runs.

This PR splits the `upload` task in two. There's now a new `prepare` task, responsible for validating local files, tracing changed files, and calculating file hashes. These steps were previously part of the `upload` task. The new `prepare` task is not skipped during dry runs.

The vast majority of the line count of this PR consists of moving step logic, UI states, tests, and stories from `upload` files to the new `prepare` files. Functional changes will be called out with inline comments.

Closes #964, closes CAP-2365

# Manual testing

Here's a build using `--only-changed`, `--trace-changed`, and `--dry-run`, showing the new UI state and the inclusion of the traced files:
![image](https://github.com/user-attachments/assets/5b232cbc-32f1-42f1-8f44-84ae20ea1156)

![image](https://github.com/user-attachments/assets/104794c6-765e-4c35-9b64-1167dd225c34)

I also ran a build without `--dry-run`, and all is working as expected.

The new UI states can also be seen in the Chromatic build for this PR.
<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>12.2.0--canary.1185.15446945773.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install chromatic@12.2.0--canary.1185.15446945773.0
  # or 
  yarn add chromatic@12.2.0--canary.1185.15446945773.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+// Get all paths in rootDir, starting at dirname.
+// We don't want the paths to include rootDir -- so if rootDir = storybook-static,
+// paths will be like iframe.html rather than storybook-static/iframe.html
+function getPathsInDirectory(ctx: Context, rootDirectory: string, dirname = '.'): PathSpec[] {
```
```suggestion
function getPathSpecsInDirectory(ctx: Context, rootDirectory: string, dirname = '.'): PathSpec[] {
```

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+  validating,
+} from '../ui/tasks/prepare';
+
+interface PathSpec {
```
I'd recommend adding a comment here indicating why a simple `Path` isn't useful. Further, do you think it would be useful to work with a member that is a true `Path`, rather than a `string` (my assumption is you chose this for some reason, so I suspect your choice is valid, I'm just curious why).

Also, thoughts on the name? I don't particularly like `PathSpec`, because it's not very descriptive, but I can't come up with another one, other than something like `PathWithContent` or `ContentfulPath`.

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+
+// Get all paths in rootDir, starting at dirname.
+// We don't want the paths to include rootDir -- so if rootDir = storybook-static,
+// paths will be like iframe.html rather than storybook-static/iframe.html
```
Since this is an API (of sorts), thoughts on making this JSDoc comment so that we can get easy descriptions on mouseover in VSCode and other tools?

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+  }
+}
+
+function getOutputDirectory(buildLog: string) {
```
Recommend adding a JSDoc comment here.

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+  return outputDirectory.trim();
+}
+
+function getFileInfo(ctx: Context, sourceDirectory: string) {
```
Recommend adding a JSDoc comment here.

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+  return { lengths, paths, statsPath, total };
+}
+
+const isValidStorybook = ({ paths, total }) =>
```
Recommend adding a JSDoc comment here, specifically on what a "valid" storybook looks like.

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+const isValidStorybook = ({ paths, total }) =>
+  total > 0 && paths.includes('iframe.html') && paths.includes('index.html');
+
+export const validateFiles = async (ctx: Context) => {
```
Recommend adding a JSDoc comment here.

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+  }
+};
+
+export const calculateFileHashes = async (ctx: Context, task: Task) => {
```
```suggestion
export async function calculateFileHashes (ctx: Context, task: Task) {
```
(to make consistent with the rest of the file)

jwir3 commented on node-src/tasks/prepare.ts:
```diff
@@ -0,0 +1,206 @@
+
+// TODO: refactor this function
+// eslint-disable-next-line complexity
+export const traceChangedFiles = async (ctx: Context, task: Task) => {
```
```suggestion
export async function traceChangedFiles (ctx: Context, task: Task) {
```

jwir3 reviewed (approved):
LGTM. I advise a couple of changes, just to maintain internal consistency within the `prepare.ts` file. @justin-thurman I know all you did was move most of the stuff over, though, so if you don't think these changes should be made, I'm ok with that. 

Your specific changes look fine.

justin-thurman commented:
@jwir3, I addressed most of your comments. I updated the function names and export syntax, and I added all the docstrings. Good calls there. I didn't make any changes to `PathSpec`, however, as I don't have any of the context there. It wasn't obvious to me, looking at the code, why it's structured that way, so I didn't want to speculate or rename it without knowing the details. Let me know what you think.

jwir3 reviewed (approved):
👍
